Fixed #26432 -- Fixed size tuple order when using numpy reshape on a GDALBand.

Opa- opened the pull request.

timgraham commented:
Could you please add a regression test in `tests/gis_tests/gdal_tests/test_raster.py`?

Opa- commented:
I improved the existing test `test_band_data` by checking the len of rows & columns of the numpy array + some arbitrary values that were swapped before.

timgraham commented:
buildbot, add to whitelist.

timgraham commented:
@yellowcap, care to review?

yellowcap commented:
Currently two `.DS_Store` files are coommited, please remove those from the patch.

yellowcap commented on django/contrib/gis/gdal/raster/band.py:
```diff
@@ -222,8 +222,10 @@ def data(self, data=None, offset=None, size=None, shape=None, as_memoryview=Fals
                 return memoryview(data_array)
             elif numpy:
+                # reshape function need the size tuple with the height first
+                reshape_size = (size[1], size[0])
```
Maybe use size.reverse() here?

Opa- replied:
size.reverse() does not exist on tuple, if you want to get something like this there is two solutions:
```python
size[::-1]
```
or
```python
tuple(reversed(size))
```
What do you think ?

yellowcap commented on django/contrib/gis/gdal/raster/band.py:
```diff
@@ -222,8 +222,10 @@ def data(self, data=None, offset=None, size=None, shape=None, as_memoryview=Fals
             if as_memoryview:
                 return memoryview(data_array)
             elif numpy:
+                # reshape function need the size tuple with the height first
```
Could reformulate to something like: "The numpy reshape function needs a reshape parameter with the height first."

Opa- commented:
@yellowcap  ok thanks for the feedbacks (didn't know about reverse, cool!)
I'll update this tomorrow :wink:

yellowcap commented on tests/gis_tests/gdal_tests/test_raster.py:
```diff
@@ -318,6 +318,19 @@ def test_band_data(self):
+            correct_list = [4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
+                            4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 3, 3, 9, 9,
+                            9, 4, 4, 4, 3, 3, 4, 4, 4, 4, 4, 4, 4, 4]
+            for x in range(0, 42):
```
Since here numpy is available, you could avoid this loop by making the correct list a numpy array and to use `numpy.testing.assert_equal`.

yellowcap commented on tests/gis_tests/gdal_tests/test_raster.py:
```diff
@@ -318,6 +318,19 @@ def test_band_data(self):
+            for x in range(0, 42):
+                self.assertEqual(data[x, 42], correct_list[x])
+            self.assertEqual(len(data), self.band.height)
+            for row in data:
```
Similarly here, you could use the `data.shape` attribute to compare directly `assertEqual(data.shape, (self.band.height, self.band.width)`.

yellowcap commented:
@Opa- you are most welcome, thanks for reporting and fixing the bug.

Opa- commented:
Hi, 

Taking your advices into account I updated the pull request, it is more clearer now thanks to you and it test the numpy array entirely. I dropped the raster.tif to txt via numpy.savetxt() and the reference array is now fully loaded & tested.

timgraham commented:
merged in 461f74ab19b7b0f393c475a55b4880227a8e4413, thanks!
